File: 01-游戏开发/07-音频/音频引擎/实时变调与变速/python/wsola.py
"""信号级链路：重采样 + WSOLA 时间拉伸（pitchshift.py 的选项位与模式约束见同目录）。"""

import math

from pitchshift import semitones_to_pitch_scale


def resample_linear(x, ratio):
    """按 ratio 改变播放速率：时长 /ratio，音高 ×ratio。"""
    n = len(x)
    m = max(1, int(round(n / ratio)))
    out = []
    for i in range(m):
        pos = i * ratio
        i0 = int(math.floor(pos))
        frac = pos - i0
        a = x[i0] if i0 < n else 0.0
        b = x[i0 + 1] if i0 + 1 < n else 0.0
        out.append(a * (1 - frac) + b * frac)
    return out


def wsola_stretch(x, ratio, frame=2048, hop=1024, search=512):
    """WSOLA 时间拉伸：时长 ×ratio，音高不变。

    与朴素 OLA 的区别是每帧先在与上一帧最相似的偏移处对齐再拼接，
    否则帧边界的相位跳变会产生可听的调制（本 demo 首版就因此把 440 Hz 估成 101 Hz）。
    """
    if len(x) < frame:
        return list(x)
    out_len = int(round(len(x) * ratio))
    out = [0.0] * out_len
    norm = [0.0] * out_len
    syn_hop = max(1, int(round(hop * ratio)))
    window = [0.5 - 0.5 * math.cos(2 * math.pi * i / frame) for i in range(frame)]
    prev = None
    read = 0
    pos = 0
    while read + frame <= len(x) and pos + frame <= out_len:
        offset = read
        if prev is not None:
            lo = max(0, read - search)
            hi = min(len(x) - frame, read + search)
            best, offset = None, read
            for off in range(lo, hi + 1, 8):
                acc = 0.0
                for k in range(0, frame, 4):
                    acc += x[off + k] * prev[k]
                if best is None or acc > best:
                    best, offset = acc, off
        block = x[offset:offset + frame]
        for i, v in enumerate(block):
            out[pos + i] += v * window[i]
            norm[pos + i] += window[i]
        prev = block
        pos += syn_hop
        read += hop
    for i in range(out_len):
        if norm[i] > 1e-9:
            out[i] /= norm[i]
    return out


# 兼容旧名
ola_stretch = wsola_stretch


def pitch_shift(x, semitones, sample_rate=48000):
    """变调不变速：先按 1/p 变速（重采样），再按 p 拉伸还原时长。"""
    p = semitones_to_pitch_scale(semitones)
    fast = resample_linear(x, p)          # 时长 /p，音高 ×p
    return ola_stretch(fast, p)           # 时长 ×p 复原，音高不变


def time_stretch(x, time_ratio):
    """变速不变调。"""
    return ola_stretch(x, time_ratio)


def estimate_frequency(x, sample_rate, min_hz=40.0, max_hz=2000.0):
    """用自相关估计基频（对 OLA 的拼接瑕疵比过零率稳健）。"""
    seg_len = min(4096, max(512, len(x) // 2))
    start = max(0, (len(x) - seg_len) // 2)
    seg = x[start:start + seg_len]
    if len(seg) < 64:
        return 0.0
    mean = sum(seg) / len(seg)
    seg = [v - mean for v in seg]
    min_lag = max(2, int(sample_rate / max_hz))
    max_lag = min(len(seg) // 2, int(sample_rate / min_hz))
    best, best_lag = 0.0, min_lag
    for lag in range(min_lag, max_lag + 1):
        acc = 0.0
        for i in range(len(seg) - lag):
            acc += seg[i] * seg[i + lag]
        if acc > best:
            best, best_lag = acc, lag
    return sample_rate / best_lag

File: 01-游戏开发/07-音频/音频引擎/实时变调与变速/go/wsola.go
package pitchshift

import "math"

// ResampleLinear 按 ratio 改变播放速率：时长 /ratio，音高 ×ratio。
func ResampleLinear(x []float64, ratio float64) []float64 {
	n := len(x)
	m := int(math.Round(float64(n) / ratio))
	if m < 1 {
		m = 1
	}
	out := make([]float64, 0, m)
	for i := 0; i < m; i++ {
		pos := float64(i) * ratio
		i0 := int(math.Floor(pos))
		frac := pos - math.Floor(pos)
		a, b := 0.0, 0.0
		if i0 < n {
			a = x[i0]
		}
		if i0+1 < n {
			b = x[i0+1]
		}
		out = append(out, a*(1-frac)+b*frac)
	}
	return out
}

// WSOLAStretch 时长 ×ratio、音高不变的波形相似重叠相加。
// 与朴素 OLA 的差别是每帧先按相似度对齐再拼接，否则帧边界的相位跳变
// 会形成可听的周期性调制（本 demo 首版即因此把 440 Hz 估成 101 Hz）。
func WSOLAStretch(x []float64, ratio float64) []float64 {
	const frame, hop, search = 2048, 1024, 512
	if len(x) < frame {
		out := make([]float64, len(x))
		copy(out, x)
		return out
	}
	outLen := int(math.Round(float64(len(x)) * ratio))
	out := make([]float64, outLen)
	norm := make([]float64, outLen)
	synHop := int(math.Round(hop * ratio))
	if synHop < 1 {
		synHop = 1
	}
	win := make([]float64, frame)
	for i := range win {
		win[i] = 0.5 - 0.5*math.Cos(2*math.Pi*float64(i)/frame)
	}
	var prev []float64
	read, pos := 0, 0
	for read+frame <= len(x) && pos+frame <= outLen {
		offset := read
		if prev != nil {
			lo, hi := read-search, read+search
			if lo < 0 {
				lo = 0
			}
			if hi > len(x)-frame {
				hi = len(x) - frame
			}
			best, bestOff := 0.0, read
			first := true
			for off := lo; off <= hi; off += 8 {
				acc := 0.0
				for k := 0; k < frame; k += 4 {
					acc += x[off+k] * prev[k]
				}
				if first || acc > best {
					best, bestOff, first = acc, off, false
				}
			}
			offset = bestOff
		}
		block := x[offset : offset+frame]
		for i, v := range block {
			out[pos+i] += v * win[i]
			norm[pos+i] += win[i]
		}
		prev = block
		pos += synHop
		read += hop
	}
	for i := range out {
		if norm[i] > 1e-9 {
			out[i] /= norm[i]
		}
	}
	return out
}

// TimeStretch 变速不变调。
func TimeStretch(x []float64, timeRatio float64) []float64 {
	return WSOLAStretch(x, timeRatio)
}

// PitchShift 变调不变速：先按 1/p 重采样，再按 p 拉伸还原时长。
func PitchShift(x []float64, semitones float64) []float64 {
	p := SemitonesToPitchScale(semitones)
	return WSOLAStretch(ResampleLinear(x, p), p)
}

// EstimateFrequency 用自相关估计基频（对拼接瑕疵比过零率稳健）。
func EstimateFrequency(x []float64, sampleRate float64) float64 {
	segLen := len(x) / 2
	if segLen > 4096 {
		segLen = 4096
	}
	if segLen < 64 {
		segLen = len(x)
	}
	if segLen < 64 {
		return 0
	}
	start := (len(x) - segLen) / 2
	seg := make([]float64, segLen)
	copy(seg, x[start:start+segLen])
	mean := 0.0
	for _, v := range seg {
		mean += v
	}
	mean /= float64(segLen)
	for i := range seg {
		seg[i] -= mean
	}
	minLag := int(sampleRate / 2000.0)
	maxLag := int(sampleRate / 40.0)
	if maxLag > segLen/2 {
		maxLag = segLen / 2
	}
	best, bestLag := 0.0, minLag
	for lag := minLag; lag <= maxLag; lag++ {
		acc := 0.0
		for i := 0; i+lag < segLen; i++ {
			acc += seg[i] * seg[i+lag]
		}
		if acc > best {
			best, bestLag = acc, lag
		}
	}
	return sampleRate / float64(bestLag)
}
